shadowfinder/shadowfinder.py

```diff
@@ -12,27 +12,47 @@
 
 class ShadowFinder:
     def __init__(
-        self, object_height=None, shadow_length=None, date_time=None, time_format="utc"
+        self,
+        object_height=None,
+        shadow_length=None,
+        date_time=None,
+        time_format="utc",
+        sun_altitude_angle=None,
     ):
 
-        self.set_details(object_height, shadow_length, date_time, time_format)
+        print(f"object_height: {object_height}")
+        print(f"shadow_length: {shadow_length}")
+        print(f"date_time: {date_time}")
+        print(f"time_format: {time_format}")
+        print(f"sun_altitude_angle: {sun_altitude_angle}")
+
+        self.set_details(
+            object_height, shadow_length, date_time, time_format, sun_altitude_angle
+        )
 
         self.lats = None
         self.lons = None
-        self.shadow_lengths = None
+        self.correct_location_probabilities = None
 
         self.timezones = None
         self.tf = TimezoneFinder(in_memory=True)
 
         self.fig = None
 
-        self.angular_resolution=0.5
-        self.min_lat=-60
-        self.max_lat=85
-        self.min_lon=-180
-        self.max_lon=180
+        self.angular_resolution = 0.5
+        self.min_lat = -60
+        self.max_lat = 85
+        self.min_lon = -180
+        self.max_lon = 180
 
-    def set_details(self, object_height, shadow_length, date_time, time_format=None):
+    def set_details(
+        self,
+        object_height,
+        shadow_length,
+        date_time,
+        time_format=None,
+        sun_altitude_angle=None,
+    ):
```

Comment:
Tbh im a bit confused, I feel like we are doing the same thing multiple times, but before that I have a question. If we have the function `__init__` that defines defaults for each input, do the defaults set by the `set_details` function not become redundant? (I'm assuming that this is the only situation where you would use `set_details`)

And then on the topic of reusability, isn't this already solved by having default values for the `__init__` function? By doing so we practically ensure that either `object_height`/`shadow_length` or `sun_altitude_angle` are being set to `None`. At that point I would have added a check before the call to `set_details` to make sure that the class is being used in one of the two modes and not in a combination.

Reply:
Ah yeah, good question 🚀 The `set_details` method is intended to be called by the user as well as by the constructor (`__init__`). [Generally the convention is that class methods that are for internal use only should start with an underscore](https://peps.python.org/pep-0008/#method-names-and-instance-variables), which isn't the case here.

The reason for wanting the user to be able to call `set_details` multiple times is that after the object is instantiated, one of `generate_timezone_grid` or `load_timezone_grid` has to be called, to populate the object with timezone data - these calls are quite expensive; `generate_timezone_grid` can take around 1 minute to run. The situation we want to avoid is the user calling this expensive method multiple times whenever they want to check a new shadow. So instead of instantiating a new object, they can reuse the existing object and call the `set_details` method before calling `find_shadows` each time.

The new situation we need to handle with the addition of the sun angle input is the user providing the details of an object and shadow on the first use of the object, and then switching to use the sun angle input on the second. At the moment, the original values for `object_height` and `shadow_length` are not cleared when `set_details` is called the second time (with `sun_altitude_angle` set) and so the if statement in `find_shadows` won't give the expected result:

https://github.com/bellingcat/ShadowFinder/blob/e998a57c63847f97c2e9ddc9183f590763042ff8/shadowfinder/shadowfinder.py#L160-L164

Does that make sense?

It might be best for the details to not be passed into the constructor at all. But that would be a different PR.

Reply:
Okay good to know, the re-usability makes sense and the use case for `set_details` also makes a lot of sense. 

As for setting either  `object_height` and `shadow_length` or `sun_altitude_angle` to `None`. I would propose something like this.

```
if not (
        (object_height is None and shadow_length is None) or 
        sun_altitude_angle is None
):
    raise ValueError("Please either set object_height and shadow_length or set sun_altitude_angle")

self.object_height = object_height
self.shadow_length = shadow_length
self.sun_altitude_angle = sun_altitude_angle
```

Since it would fail explicitly if the user of the function calls the function with unclear parameters. What I mean is that in the case where the user might set all three, the function fails, avoiding unclear situations.

Reply:
I had to draw a truth table to wrap my head around the logic here! 😂 

I think it also needs to handle the case where a user incorrectly gives one of the length-inputs and not the other.

This is the most concise Python expression that makes some intuitive sense to me.

```python
valid_input = (object_height is None == shadow_length is None) and # height and length must have the same None-ness
              (object_height is None ^ sun_altitude_angle is None) # either height or angle must be set (but not both or neither)
              
if not valid_input:
    raise ValueError("Please either set object_height and shadow_length or set sun_altitude_angle")
```
I think the brackets are needed, but they definitely help understanding.

Feel free to try coming up with something else, but also feel free to use that one!

Reply:
Oh yes, definitely did not think of that edge case and that code looks great! Love the use of the xor. But it took me waay to long to realize that even more brackets are needed otherwise the `object_height is None` is evaluated and then the result of that compared to `shadow_length` instead of doing the two `is None`s and then the `==`.